pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -125,14 +237,255 @@ func localityLbEndpointsFromInstances(instances []*model.ServiceInstance) []endp
+			if len(con.Clusters) == 0 {
+				continue
+			}
+		}
+
+		if len(con.Clusters) > 0 {
+			response := s.endpoints(s.mesh, con.Clusters)
+			err := stream.Send(response)
+			if err != nil {
+				return err
+			}
+
+			if edsDebug {
+				log.Infof("EDS RES for %s %q clusters %v, Response: \n%s\n",
+					node, peerAddr, con.Clusters, response.String())
+			}
+		} else {
+			if edsDebug {
+				log.Infof("EDS empty clusters %v \n", peerAddr)
+			}
+		}
+	}
+}
+
+// EdsPushAll implements old style invalidation, generated when any rule or endpoint changes.
+// Primary code path is from v1 discoveryService.clearCache(), which is added as a handler
+// to the model ConfigStorageCache and Controller.
+func EdsPushAll() {
+	if edsDebug {
+		log.Infoa("EDS cache reset")
+	}
+	edsClusterMutex.Lock()
+	// Create a temp map to avoid locking the add/remove
+	tmpMap := map[string]*EdsCluster{}
+	for k, v := range edsClusters {
+		tmpMap[k] = v
+	}
+	version++
+	edsClusterMutex.Unlock()
+
+	for clusterName, edsCluster := range tmpMap {
+		updateCluster(clusterName, edsCluster)
+		for _, edsCon := range edsCluster.EdsClients {
+			edsCon.pushChannel <- true
+		}
+	}
+}
+
+// Edsz implements a status and debug interface for EDS.
+// It is mapped to /debug/edsz on the monitor port (9093).
+func Edsz(w http.ResponseWriter, req *http.Request) {
+	if req.Form.Get("debug") != "" {
+		edsDebug = req.Form.Get("debug") == "1"
+		return
+	}
+	if req.Form.Get("push") != "" {
+		EdsPushAll()
+	}
+	data, err := json.Marshal(edsClusters)
+	if err != nil {
+		_, _ = w.Write([]byte(err.Error()))
+		return
+	}
+
+	_, _ = w.Write(data)
+}
+
+// addEdsCon will track the eds connection, for push and debug
+func (s *DiscoveryServer) addEdsCon(clusterName string, node string, connection *EdsConnection) {
+
+	c := s.getOrAddEdsCluster(clusterName)
+	c.mutex.Lock()
+	existing := c.EdsClients[node]
+	c.mutex.Unlock()
+
+	// May replace an existing connection
+	if existing != nil {
+		existing.pushChannel <- false // force closing it
+	}
+	c.EdsClients[node] = connection
+}
+
+// getEdsCluster returns a cluster.
+func (s *DiscoveryServer) getEdsCluster(clusterName string) *EdsCluster {
+	// separate method only to have proper lock.
+	edsClusterMutex.Lock()
+	defer edsClusterMutex.Unlock()
+	return edsClusters[clusterName]
+}
+
+func (s *DiscoveryServer) getOrAddEdsCluster(clusterName string) *EdsCluster {
+	edsClusterMutex.Lock()
+	defer edsClusterMutex.Unlock()
+
+	c := edsClusters[clusterName]
+	if c == nil {
+		c = &EdsCluster{discovery: s,
+			EdsClients: map[string]*EdsConnection{},
+			FirstUse:   time.Now(),
+		}
+		edsClusters[clusterName] = c
+	}
+	return c
+}
+
+// removeEdsCon is called when a gRPC stream is closed, for each cluster that was watched by the
+// stream. As of 0.7 envoy watches a single cluster per gprc stream.
+func (s *DiscoveryServer) removeEdsCon(clusterName string, node string, connection *EdsConnection) {
+	c := s.getEdsCluster(clusterName)
```

Comment:
you could do the check in this function and avoid multiple lock/unlock cycles

Reply:
will try it out in separate PR - not sure.